### PR TITLE
use typescript

### LINKED ISSUE #127: use TypeScript

### PR DESCRIPTION
**Pull request by shoota**
Closes #127

### CODE REVIEW

**Comment by shoota** [comment]
WIP.
Only `src` use typescript.

**Comment by shoota** [comment]
WIP. test using ts, but test running with transpiled js.

**Comment by shoota** [comment]
`karma-typescript`でloadさせると何故か`Object.assign`がないと言われてしまう

**Comment by shoota** [comment]
rebased

**Comment by shoota** [comment]
sauceでコケる
```
ERROR [karma]: TypeError: Cannot read property 'writeQueueSize' of null
 at Socket._onTimeout (net.js:401:43)
 at ontimeout (timers.js:471:11)
 at tryOnTimeout (timers.js:306:5)
 at Timer.listOnTimeout (timers.js:266:5)
```

File: package.json [hunk 1]
```diff
@@ -69,9 +69,8 @@
   ],
   "license": "GPL-3.0",
   "lint-staged": {
-    "*.js": [
-      "prettier-eslint --write",
```
Comment (9renpoto) [inline package.json]: why disable prettier

File: package.json [hunk 2]
```diff
@@ -69,9 +69,8 @@
...
-      "eslint --cache",
+    "*.{ts,tsx}": [
```
Comment (9renpoto) [inline package.json]: tsx is not used

File: package.json [hunk 3]
```diff
@@ -82,19 +81,20 @@
   },
   "scripts": {
     "build": "run-p build:*",
-    "build:babel": "cross-env NODE_ENV=dev babel src -d lib",
+    "build:ts": "cross-env NODE_ENV=dev tsc",
```
Comment (9renpoto) [inline package.json]: maybe unused NODE_ENV when `tsc`

File: karma.conf.ts [hunk 4]
```diff
@@ -65,7 +70,7 @@ if (
   )
 }
 
-const setting = Object.assign({}, base, override)
+const setting = objectAssign({}, base, override)
```
Comment (9renpoto) [inline karma.conf.ts]: `Object.assign({}, base, override)` is work, because `karma.conf.(js|ts)` works nodejs not browser .

Comment (shoota) [reply karma.conf.ts]: oh, sorry, My thinking didn't enough. thanks.

**Comment by shoota** [comment]
rebased

File: package.json [hunk 5]
```diff
@@ -7,34 +7,30 @@
...
-    "babel-plugin-transform-object-assign": "^6.22.0",
-    "babel-preset-env": "^1.6.0",
-    "babel-preset-flow": "^6.23.0",
-    "babel-preset-power-assert": "^1.0.0",
-    "babel-register": "^6.24.1",
-    "codecov": "^3.0.0",
+    "codecov": "^2.3.0",
     "cross-env": "^5.0.5",
     "esdoc": "^1.0.1",
     "esdoc-flow-type-plugin": "^1.0.0",
```
Comment (9renpoto) [inline package.json]: is this need?

File: package.json [hunk 6]
```diff
@@ -47,11 +43,16 @@
     "lint-staged": "^4.0.3",
     "mocha": "^4.0.0",
     "npm-run-all": "^4.0.2",
+    "object-assign": "^4.1.1",
```
Comment (9renpoto) [inline package.json]: not `devDependencies`, because include production

File: package.json [hunk 7]
```diff
@@ -7,34 +7,30 @@
...
-    "babel-core": "^6.25.0",
-    "babel-loader": "^7.1.1",
-    "babel-plugin-istanbul": "^4.1.4",
-    "babel-plugin-transform-object-assign": "^6.22.0",
-    "babel-preset-env": "^1.6.0",
-    "babel-preset-flow": "^6.23.0",
-    "babel-preset-power-assert": "^1.0.0",
-    "babel-register": "^6.24.1",
-    "codecov": "^3.0.0",
+    "codecov": "^2.3.0",
```
Comment (9renpoto) [inline package.json]: why rollback

File: src/core.ts [hunk 8]
```diff
@@ -112,13 +118,13 @@ export default class AgentCore extends Store {
     }
   }
 
-  _updateInteractCache (data: Object): void {
+  _updateInteractCache (data: Interact): void {
```
Comment (9renpoto) [inline src/core.ts]: You did not have to change this part. Because it have validator

File: src/core.ts [hunk 9]
```diff
@@ -128,10 +134,13 @@ export default class AgentCore extends Store {
     })
 
     if (this._baseUrl && (query.length >= MAX_INTERACT || force)) {
+      const customState: any = this.get('custom')
```
Comment (9renpoto) [inline src/core.ts]: `s/any/Object/` ?

File: src/global.ts [hunk 10]
```diff
@@ -22,11 +25,11 @@ export default function (Agent: any) {
     return agents[name][api[0]](...tasks)
   }
 
-  ;((global: window) => {
+  ((global: any) => {
     const element: any = document.querySelector(`[${NAMESPACE}]`)
     const name: string = element.getAttribute(NAMESPACE)
     if (global[name] && global[name].q) {
-      const queue = (global[name].q: Array<TaskQueue>)
+      const queue: TaskQueue = global[name].q
```
Comment (9renpoto) [inline src/global.ts]: task queue it not `Array`
`Array<TaskQueue>`

File: webpack.config.ts [hunk 11]
```diff
@@ -1,27 +1,29 @@
-/* @flow */
 const webpack = require('webpack')
 const UglifyJSPlugin = require('uglifyjs-webpack-plugin')
 const path = require('path')
-const { version, name, author, license } = require('./package.json')
+const { version, moduleName, author, license } = require('./package.json')
```
Comment (9renpoto) [inline webpack.config.ts]: `moduleName` is this defined ?
Im not fount it.

Comment (9renpoto) [reply webpack.config.ts]: `/*! undefined 0.9.2 | Copyright (c) 2017 UNCOVER TRUTH Inc. | License GPL-3.0 */`

File: webpack.config.ts [hunk 12]
```diff
@@ -1,27 +1,30 @@
-/* @flow */
 const webpack = require('webpack')
 const UglifyJSPlugin = require('uglifyjs-webpack-plugin')
 const path = require('path')
-const { version, name, author, license } = require('./package.json')
+const { version, author, license } = require('./package.json')
+const packageName = '@userdive/agent'
```
Comment (shoota) [inline webpack.config.ts]: if use `name` occur ts error, `cannot redeclare block scoped variable (typescript)` ... 😢 
And I have no idea...

File: src/core.ts (hunk 8 above)
Comment (shoota) [reply src/core.ts]: validator function argument needs typing causes `noImplicitAny` compile. 
I guess that the arg type should be 'Interact' not `Object`,  and `_updateInteractCache` arg 
 type should be same.

File: src/core.ts (hunk 9 above)
Comment (shoota) [reply src/core.ts]: `customeState` will use to `obj2query`. the function need argument Object __value typing__ for accept string function, `encodeURIComponent`. 
So Object can not assign `obj2query` arg.

**Comment by shoota** [comment]
docs generate by `typedocs`. plz run `build` and check it too.

File: package.json [hunk 13]
```diff
@@ -82,19 +84,20 @@
...
-    "lint": "run-p lint:*",
-    "lint:es": "eslint . --cache",
-    "lint:flow": "flow",
+    "lint": "tslint \"src/**/*.ts\" \"test/**/*.ts\"",
     "postbuild": "cp robots.txt lib/build/.",
     "prebuild": "npm run clean",
     "precommit": "lint-staged",
     "prepublish": "npm run build",
-    "test": "karma start --single-run"
+    "test": "run-s test:*",
```
Comment (9renpoto) [inline package.json]: maybe, can use `run-p`

File: src/requests.ts [hunk 14]
```diff
@@ -12,21 +11,21 @@ export function enable (): boolean {
...
     img.onload = onload
     img.onerror = onerror
     img.src = `${url}?${query.join('&')}`
   }
 }
 
-export function obj2query (data: Object): string[] {
-  const query = []
+export function obj2query (data: { key: any; uriComponent: string }): string[] {
+  const query: any[] = []
```
Comment (9renpoto) [inline src/requests.ts]: query is `string[]` ?

**Comment by 9renpoto** [comment]
try this ? https://github.com/power-assert-js/espower-typescript

**Pull request merged**